make docstring-update optional for io.registry and some additional performance changes

MSeifert04 opened the pull request:
This further (see #5262 for links and rationale) decreases the creation time of the docstring-tables for `read` and `write` functions using the `io.registry`:

## ``_is_best_match``
This function was responsible for 10% of the time spent in `get_formats`. The time was "used" to create the list of possible classes but this list was only necessary _if_ `class1 is not class2`. This check is now done before creating the list.

The second improvement here (maybe it's currently rather a disadvantage) is that instead of creating a list a `set` is created. This (a) speeds up the `not in` comparison, (b) discards duplicates but (c) is slower than creating a list and the `in` operation is also slower because the list only contained 36 `Table`-links (a speedup would only be visible if the list would contain several distinct classes)

With these changes the relative time spent in `_is_best_match` was reduced to 3%. This is not really noticable though because other parts take much more time. In effect a call to `register_reader` or `register_writer` is sped up by 2%.

## ``get_formats`` and ``_update__doc__``
20% of the time spent in `_update__doc__` was because of the line `format_table = format_table[has_readwrite]`. By replacing this with a list comprehension in `get_formats` these 20% were brought down to less than 1%.

This change noticably reduced the `import astropy.coordinates` time (and similar subpackages which import Table) by 3-4%

## Making the ``_update__doc__`` function optional

As alternative to #5264 which proposed to create the docstring dynamically this PR includes a change which makes the ``_update__doc__`` function optional. The registration of readers/writers don't force the update anymore (but do it by default for backwards compatibility) and it's disabled for all in-astropy-registrations. However after importing `astropy.table` the docs are created _once_.

## One additional change

Instead of doing `lines.extend([' ' * left_indent + line for line in new_lines])` the `' ' * left_indent` is determined before the list-comprehension. This yields an amazing 0.01% speedup.

## Result

The `python -m timeit -n 1 -r 1 "from astropy.coordinates import SkyCoord"` command takes 13% less time than before. (At least on my computer :smile:)

@mhvk Since you proposed the dynamic creation maybe you can judge if this approach would also fix your issue (or if such a change should be avoided).

MSeifert04 commented:
Anyone interested in doing a review?

pllim commented:
Since there is some API change, I think it warrants a change log, no? Also, I think @taldcroft might be the most qualified to review this.

mhvk commented on astropy/io/registry.py:
```diff
@@ -77,6 +82,15 @@ def get_formats(data_class=None):
+            rows = [row for row in rows if row[3] == 'Yes']
+        else:
+            raise ValueError('unrecognized value for "readwrite": {0}.\n'
+                             'Allowed are "Read" and "Write".')
```
I think this should be: 'Allowed are "Read", "Write", and None.'

MSeifert04 replied:
Thanks, I've changed it.

mhvk commented on astropy/io/registry.py:
```diff
@@ -414,10 +432,11 @@ def _is_best_match(class1, class2, format_classes):
+    # class1 is a subclass of class2.
+    return (class1 is class2 or
+            (issubclass(class1, class2) and
+             class1 not in {cls for fmt, cls in format_classes}))
```
I thought straight iterator with `()` might be faster yet, but a quick test on my python3 shows that is not true; this does seem the fastest...

MSeifert04 replied:
It really depends on what `class1` is and if it's present in the `format_classes` and how big the `format_classes` are.

But the biggest improvement here is that the set is created only if the other conditions require it.

mhvk commented on astropy/table/__init__.py:
```diff
@@ -39,3 +39,8 @@ class Conf(_config.ConfigNamespace):
 from . import jsviewer
+
+from ..io import registry
+
```
Would it be an idea to comment here that the readers/writers register automatically, and this completes it by updating the documentation?

MSeifert04 replied:
I agree. I've added a short comment there.

mhvk commented:
@MSeifert04 - I think this is very nice, and less complicated than my suggestion.

taldcroft commented:
@MSeifert04 - thanks for looking in to this.  Do you have a feeling for how much improvement comes from the code change to make `_update_doc_` be optional?  My main question is changing the API and adding complexity if there is not a substantial practical gain.

MSeifert04 commented:
@taldcroft I'm glad you asked.

```
from astropy.io import registry
from astropy.table import Table

def func():
    pass

%timeit registry.register_reader('blub', Table, func, force=True)

# 100 loops, best of 3: 11.2 ms per loop  (before)
# 100 loops, best of 3: 7.89 ms per loop  (after)
```

Currently the `_update_doc_` function is called 68 times on startup, with this PR it is called 2 times. So this is an improvement of rougly 700ms total (and still 500ms if I removed the optionality of `_update_doc_` from the PR).

```
# with PR
$ python -m timeit -n 1 -r 1 "from astropy.coordinates import SkyCoord"
1 loops, best of 1: 4.24 sec per loop

# without PR
$ python -m timeit -n 1 -r 1 "from astropy.coordinates import SkyCoord"
1 loops, best of 1: 5.08 sec per loop
```

Generally "modern" computers are 3-4 times faster than my old laptop, so having another data-point would be good.

taldcroft commented:
OK, so it sounds like the improvement from the optionality alone is around 200 msec, or about 4%.  Is that right?  BTW on my machine (oldish iMac) my import times were very nearly a factor of 2 smaller than yours.

So 4% is not huge.. but what about doing the change in a way that doesn't require an API change?  I wonder if it would work to put a global set within `registry.py` called (e.g.) `_update_docs`.  This would initially be empty and you would have code like below within `register_reader/writer`:
```
if data_class in _update_docs:
    _update__doc__(data_class, 'read')
```
Then within `_update__doc__` the first thing would be:
```
# From now on any reader/writer registrations will automatically update
# docs at the time of registration.  By default this does not happen in order
# to defer these updates to a single update at the end of imports.  Note that
# this requires each data_class that uses the I/O registry to explicitly call
# _update_docs as soon as the class is exposed (e.g. in the subpackage __init__.py).
_update_docs.add(data_class)
```

BTW, since `_update__doc__` is effectively now public (with either implementation) it should have the leading underscore removed and be treated as a public method.

MSeifert04 commented:
@taldcroft Sorry, I realize I was a bit unclear in https://github.com/astropy/astropy/pull/5275#issuecomment-245012691. It's 700ms faster just by making it optional. The 200ms difference is the improvement made by the other changes (if one wouldn't make the `update_doc` optional). So **without making it optional** I can offer 4% improvement and with it it's approximatly 15% faster. However I would be very happy if someone else could verify this number.

MSeifert04 commented:
@taldcroft I'm not sure if I understand your suggestion. Too many `_update_doc`, `_update__doc__` variables. Do you mean instead of a parameter some value that is `False` that will be changed to `True` as soon as the astropy-subpackage is done with it?

taldcroft commented:
>  Do you mean instead of a parameter some value that is False that will be changed to True as soon as the astropy-subpackage is done with it?

Roughly, but where the True/False is defined by whether `data_class` is in a global `set`.  Below is the prescription from above but with a more unique name for the global set.

I wonder if it would work to put a global set within `registry.py` called (e.g.) `auto_update_docs`.  This would initially be empty and you would have code like below within `register_reader/writer`:
```
if data_class in auto_update_docs:
    _update__doc__(data_class, 'read')
```
Then within `_update__doc__` the first thing would be:
```
# By adding ``data_class` to the ``auto_update_docs`` set, from now on 
# any reader/writer registrations will automatically update
# docs at the time of registration.  By default this does not happen in order
# to defer these updates to a single update at the end of imports.  Note that
# this requires each data_class that uses the I/O registry to explicitly call
# _update_docs as soon as the class is exposed (e.g. in the subpackage __init__.py).

auto_update_docs.add(data_class)
```

BTW, since `_update__doc__` is effectively now public (with either implementation) it should have the leading underscore removed and be treated as a public method.

MSeifert04 commented:
@taldcroft thank you. Now I understand it!

It probably would work and likely yield the same import-time increase. What I'm not sure about which approach is better. :smile: I'd rather go with the added parameter because it wouldn't add a mutable public global variable and it's likely not backwards compatible. For example if I had a class that has some reader/writer I'd need to add it to the `auto_update_docs` and/or call `_update__doc__` which wasn't necessary before and isn't necessary with the approach in this PR.

But I'll think about it some more.

taldcroft commented:
The point is that this new parameter is really not functionally useful except as a way to reduce import times.  So if we later think of a different / better way to manage import times, for instance #3090, then we are stuck with a useless parameter in the API.

mhvk commented:
I think I agree with @taldcroft here: it is good to focus on the specific goal for now of doing just one update after we know all normal astropy classes have initialized, and if we can do that without introducing any new outside-visible options, that would be great. Though I would perhaps take that to even further extent: of setting some `_donot_update_doc_yet` variable at the start, and removing that at the end of initialization.

taldcroft commented:
BTW, I did not intend for the the global in my approach to be public, that was just a mistake in the previous comment.  I have edited it accordingly.

One could have back compatibility by explicitly turning doc updates on and off via a public method.  So essentially only classes like `Table` where this is a real problem would bother:
```
from ..io import registry

registry.disable_auto_doc_updates(Table)

from ..io.ascii import connect
from ..io.fits import connect
from ..io.misc import connect
from ..io.votable import connect
from . import jsviewer

registry.enable_auto_doc_updates(Table)
```
The `enable_auto_doc_updates` function would also actually do the update as well, but by default auto updating would be enabled.  Thus the module global would be the set of  `data_class` for which auto updating is disabled.

MSeifert04 commented:
I don't see the advantage of adding another public function (or two) instead of adding a parameter. As far as I see the `register_reader`, etc. functions could be deprecated as soon as the code is overhauled and then the additional parameter wouldn't matter.

MSeifert04 commented:
Even though I still think having a deprecated parameter would be more senseable than creating a new function or functionality I've tried to implement some combination of @mhvk and @taldcroft suggestion and it could work with a context manager (unnecessary details are left out):

```
from astropy.table import Table

_donot_update_docs_yet = set() 

def _update__doc__(cls, func):
    print('created docs for {}.{}'.format(cls, func))
    
def register_reader(cls):
    print('registered reader')
    if cls not in _donot_update_docs_yet:
        _update__doc__(cls, 'read')
    
def register_writer(cls):
    print('registered writer')
    if cls not in _donot_update_docs_yet:
        _update__doc__(cls, 'write')

class PostponeDocCreation(object):
    def __init__(self, cls):
        self._cls = cls
        
    def __enter__(self):
        _donot_update_docs_yet.add(self._cls)
        
    def __exit__(self, *args):
        _update__doc__(self._cls, 'read')
        _update__doc__(self._cls, 'write')
        _donot_update_docs_yet.discard(self._cls)
```

which would be used like this:

```
with PostponeDocCreation(Table):
    register_reader(Table)
    register_writer(Table)
    register_reader(Table)
    register_writer(Table)
    register_reader(Table)
    register_writer(Table)
    register_reader(Table)
    register_writer(Table)

#########
registered reader
registered writer
registered reader
registered writer
registered reader
registered writer
registered reader
registered writer
created docs for <class 'astropy.table.table.Table'>.read
created docs for <class 'astropy.table.table.Table'>.write
```

If you think a new functionality is better than a new parameter I can implement this. I think I already made it clear that I prefer the additional parameter :smile:

taldcroft commented:
The context manager idea looks fine, though I would propose using `contextlib.contextmanager` to make the code simpler and easier to read.

MSeifert04 commented:
@taldcroft I've commited the contextmanager.

taldcroft commented on astropy/io/registry.py:
```diff
@@ -41,7 +43,49 @@ class IORegistryError(Exception):
+@contextlib.contextmanager
+def delay_doc_updates(cls):
+    """Contextmanager to disable documentation updates when registering
+    reader and writer. The documentation is only build once when the
```
typo `build` => `built`

taldcroft commented on astropy/io/registry.py:
```diff
@@ -41,7 +43,49 @@ class IORegistryError(Exception):
+
+    Notes
+    -----
+    Registering mutliple reader and writer can cause significant overhead
```
`multiple readers and writers`

taldcroft commented on astropy/io/registry.py:
```diff
@@ -41,7 +43,49 @@ class IORegistryError(Exception):
+    -----
+    Registering mutliple reader and writer can cause significant overhead
+    because the documentation of the corresponding ``read`` and ``write``
+    method are build every time.
```
`methods`

taldcroft commented:
@MSeifert04 - apart from the minor doc comments this looks fine, thanks!

MSeifert04 commented on CHANGES.rst:
```diff
@@ -92,6 +92,9 @@ API Changes
   - ``.fts`` and ``.fts.gz`` files will be automatically identified as
     ``io.fits`` files if no explicit ``format`` is given. [#5211]
 
+  - ``delay_doc_updates`` contextmanager to postpone the formatting of the
```
Should this be in the API section or is it a new feature?

MSeifert04 replied:
and should the import time improvement be mentioned in the "Others" section?

mhvk replied:
I'd say it is a new feature, but I'm not sure we really want to expose it as such, so my feeling would be in "Others" section only.

mhvk commented on astropy/io/registry.py:
```diff
@@ -15,7 +17,7 @@
 __all__ = ['register_reader', 'register_writer', 'register_identifier',
            'identify_format', 'get_reader', 'get_writer', 'read', 'write',
-           'get_formats', 'IORegistryError']
+           'get_formats', 'IORegistryError', 'delay_doc_updates']
```
Should this be exported at all? We only use it directly as `registry.delay_doc_updates`?

mhvk commented:
@MSeifert04 - looks very good.  I only wonder if we shouldn't keep it sort-of private for now by not putting it in `__all__` and not mentioning it as a new feature, but just under the speed-up note. But as I do quite like the idea, I'm also happy to put it out there (though then in "new feature" rather than API change).

MSeifert04 commented:
@mhvk Good question. I don't know.

taldcroft commented:
I don't have a strong opinion about the publicity of this.  Either way...

MSeifert04 commented:
I've moved the changelog entry to the "Other changes" section but left it inside ``__all__``. The primary motivation was the speedup when importing astropy-subpackages and it has a big warning box inside the docstring.

There was some weird merge conflict message (changelog related) so I rebased this and restarted the tests.

taldcroft commented:
OK, just one last thing, please start both of the changelog entries with `Added ...` for stylistic consistency (they are capitalized and have a verb).  This can be a `[skip ci]` commit.  Also the second one needs a blank line above it.

MSeifert04 commented:
@taldcroft done.

taldcroft commented:
Thanks @MSeifert04 !
